src/ray/gcs/gcs_server/gcs_heartbeat_manager.cc

```diff
@@ -0,0 +1,111 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/gcs/gcs_server/gcs_heartbeat_manager.h"
+#include "ray/common/ray_config.h"
+#include "ray/gcs/pb_util.h"
+#include "src/ray/protobuf/gcs.pb.h"
+
+namespace ray {
+namespace gcs {
+
+GcsHeartbeatManager::GcsHeartbeatManager(
+    boost::asio::io_service &io_service,
+    std::function<void(const NodeID &)> on_node_death_callback)
+    : io_service_(io_service),
+      on_node_death_callback_(std::move(on_node_death_callback)),
+      num_heartbeats_timeout_(RayConfig::instance().num_heartbeats_timeout()),
+      detect_timer_(io_service) {
+  io_service_thread_.reset(new std::thread([this] {
+    /// The asio work to keep io_service_ alive.
+    boost::asio::io_service::work io_service_work_(io_service_);
+    io_service_.run();
+  }));
+}
+
+void GcsHeartbeatManager::Start() {
+  io_service_.post([this] {
+    if (!is_started_) {
+      Tick();
+      is_started_ = true;
+    }
+  });
+}
+
+void GcsHeartbeatManager::Stop() {
+  io_service_.stop();
+  if (io_service_thread_->joinable()) {
+    io_service_thread_->join();
+  }
+}
+
+void GcsHeartbeatManager::AddNode(const NodeID &node_id) {
+  io_service_.post(
+      [this, node_id] { heartbeats_.emplace(node_id, num_heartbeats_timeout_); });
+}
+
+void GcsHeartbeatManager::HandleReportHeartbeat(
+    const rpc::ReportHeartbeatRequest &request, rpc::ReportHeartbeatReply *reply,
+    rpc::SendReplyCallback send_reply_callback) {
+  NodeID node_id = NodeID::FromBinary(request.heartbeat().node_id());
+  auto iter = heartbeats_.find(node_id);
+  if (iter == heartbeats_.end()) {
+    // Ignore this heartbeat as the node is not registered.
+    // TODO(Shanly): Maybe we should reply the raylet with an error. So the raylet can
+    // crash itself as soon as possible.
+    return;
+  }
+
+  iter->second = num_heartbeats_timeout_;
+  GCS_RPC_SEND_REPLY(send_reply_callback, reply, Status::OK());
+}
+
+/// A periodic timer that checks for timed out clients.
+void GcsHeartbeatManager::Tick() {
+  DetectDeadNodes();
+  ScheduleTick();
+}
+
+void GcsHeartbeatManager::DetectDeadNodes() {
+  for (auto it = heartbeats_.begin(); it != heartbeats_.end();) {
+    auto current = it++;
+    current->second = current->second - 1;
+    if (current->second == 0) {
+      auto node_id = current->first;
+      RAY_LOG(WARNING) << "Node timed out: " << node_id;
+      heartbeats_.erase(current);
+      if (on_node_death_callback_) {
+        on_node_death_callback_(node_id);
+      }
+    }
+  }
+}
+
+void GcsHeartbeatManager::ScheduleTick() {
+  auto heartbeat_period = boost::posix_time::milliseconds(
+      RayConfig::instance().raylet_heartbeat_timeout_milliseconds());
```

Comment:
Should we increase the heartbeat update time? I think it is less crucial compared to resource update time.

Reply:
yes. Internally we changed it to 1 seconds. I'll do the same change here.